src/lightning/app/core/app.py

```diff
@@ -216,76 +217,79 @@ def get_component_by_name(self, component_name: str):
 
         current = self.root
         for child_name in component_name.split(".")[1:]:
-            if isinstance(current, (LightningDict, LightningList)):
-                child = current[child_name] if isinstance(current, LightningDict) else current[int(child_name)]
+            if isinstance(current, LightningDict):
+                child = current[child_name]
+            elif isinstance(current, LightningList):
+                child = current[int(child_name)]
             else:
                 child = getattr(current, child_name, None)
             if not isinstance(child, ComponentTuple):
                 raise AttributeError(f"Component '{current.name}' has no child component with name '{child_name}'.")
-            current = child
+            current = child  # type: ignore[assignment]
```

Comment:
you can probably add the type to `current` or `self.root` to resolve this issue

Reply:
Yes, but then the return type does not match and would need some more asserts... anyway this loop is dirty, so let's leave it as preserving the actual state fix by in another work... :innocent: